otherlibs/systhreads/st_win32.h

```diff
@@ -160,53 +174,100 @@ Caml_inline void st_thread_yield(st_masterlock * m)
+     accessed without holding "crit", using the Interlocked API for
+     atomic accesses */
+};
+
+typedef struct st_mutex_ * st_mutex;
 
 static DWORD st_mutex_create(st_mutex * res)
 {
-  st_mutex m = caml_stat_alloc_noexc(sizeof(CRITICAL_SECTION));
+  st_mutex m = caml_stat_alloc_noexc(sizeof(struct st_mutex_));
   if (m == NULL) return ERROR_NOT_ENOUGH_MEMORY;
-  InitializeCriticalSection(m);
+  InitializeCriticalSection(&m->crit);
+  m->owner = 0;
   *res = m;
   return 0;
 }
 
 static DWORD st_mutex_destroy(st_mutex m)
 {
-  DeleteCriticalSection(m);
+  DeleteCriticalSection(&m->crit);
   caml_stat_free(m);
   return 0;
 }
 
+/* Error codes with the 29th bit set are reserved for the application */
+
+#define MUTEX_DEADLOCK (1<<29 | 1)
+#define MUTEX_PREVIOUSLY_UNLOCKED 0
+#define MUTEX_ALREADY_LOCKED (1 << 29)
+#define MUTEX_NOT_OWNED (1<<29 | 2)
+
 Caml_inline DWORD st_mutex_lock(st_mutex m)
 {
+  st_tid self, prev;
   TRACE1("st_mutex_lock", m);
-  EnterCriticalSection(m);
+  self = st_current_thread_id();
+  /* Critical sections are recursive locks, so this will succeed
+     if we already own the lock */
+  EnterCriticalSection(&m->crit);
+  /* Record that we are the owner of the lock */
+  prev = Tid_Atomic_Exchange(&m->owner, self);
+  if (prev != 0) {
+    /* The mutex was already locked by ourselves.
+       Cancel the EnterCriticalSection above and return an error. */
+    TRACE1("st_mutex_lock (deadlock)", m);
+    LeaveCriticalSection(&m->crit);
+    return MUTEX_DEADLOCK;
+  }
   TRACE1("st_mutex_lock (done)", m);
   return 0;
 }
 
-/* Error codes with the 29th bit set are reserved for the application */
-
-#define PREVIOUSLY_UNLOCKED 0
-#define ALREADY_LOCKED (1<<29)
-
 Caml_inline DWORD st_mutex_trylock(st_mutex m)
 {
+  st_tid self, prev;
   TRACE1("st_mutex_trylock", m);
-  if (TryEnterCriticalSection(m)) {
-    TRACE1("st_mutex_trylock (success)", m);
-    return PREVIOUSLY_UNLOCKED;
-  } else {
+  self = st_current_thread_id();
+  if (! TryEnterCriticalSection(&m->crit)) {
     TRACE1("st_mutex_trylock (failure)", m);
-    return ALREADY_LOCKED;
+    return MUTEX_ALREADY_LOCKED;
+  }
+  /* Record that we are the owner of the lock */
+  prev = Tid_Atomic_Exchange(&m->owner, self);
+  if (prev != 0) {
+    /* The mutex was already locked by ourselves.
+       Cancel the EnterCriticalSection above and return "already locked". */
+    TRACE1("st_mutex_trylock (already locked by self)", m);
+    LeaveCriticalSection(&m->crit);
+    return MUTEX_ALREADY_LOCKED;
   }
+  TRACE1("st_mutex_trylock (done)", m);
+  return MUTEX_PREVIOUSLY_UNLOCKED;
 }
 
 Caml_inline DWORD st_mutex_unlock(st_mutex m)
 {
+  st_tid self, prev;
+  /* If the calling thread holds the lock, m->owner is stable and equal
+     to GetCurrentThreadId().
+     Otherwise, the value of m->owner can be 0 (if the mutex is unlocked)
+     or some other thread ID (if the mutex is held by another thread),
+     but is never equal to GetCurrentThreadId(). */
```

Comment:
```suggestion
  /* If the calling thread holds the lock, m->owner is stable and equal
     to st_current_thread_id().
     Otherwise, the value of m->owner can be 0 (if the mutex is unlocked)
     or some other thread ID (if the mutex is held by another thread),
     but is never equal to st_current_thread_id(). */
```